src/iter/collect/consumer.rs

```diff
@@ -70,44 +112,74 @@ impl<'c, T: Send + 'c> Consumer<T> for CollectConsumer<'c, T> {
 }
 
 impl<'c, T: Send + 'c> Folder<T> for CollectFolder<'c, T> {
-    type Result = ();
+    type Result = PartialCollect<T>;
 
     fn consume(mut self, item: T) -> CollectFolder<'c, T> {
-        // Compute target pointer and write to it. Safe because the iterator
-        // does all the bounds checking; we're only avoiding the target drop.
-        let head = self
-            .target
-            .next()
-            .expect("too many values pushed to consumer");
+        // Compute target pointer and write to it
+
+        // WIP: Does not work yet if T is ZST
+        if self.ptr == self.end {
+            panic!("too many values pushed to consumer");
+        }
         unsafe {
+            let head = self.ptr;
+            self.ptr = self.ptr.add(1);
             ptr::write(head, item);
         }
 
         self.local_writes += 1;
```

Comment:
`ptr` and `local_writes` are counting the same sequence in parallel, so that can be updated